src/value_generators/src/DocumentGenerator.cpp

```diff
@@ -561,6 +872,85 @@ UniqueGenerator<int64_t> intGenerator(const Node& node, DefaultRandom& rng) {
+ *   or a constant generator if given a constant/scalar.
+ */
+UniqueGenerator<double> doubleGenerator(const Node& node, DefaultRandom& rng) {
+    // Set of parsers to look when we request an double parser
+    // see doubleGenerator
+    const static std::map<std::string, Parser<UniqueGenerator<double>>> doubleParsers{
+        {"^RandomDouble", doubleGeneratorBasedOnDistribution},
+    };
+
+    if (auto parserPair = extractKnownParser(node, rng, doubleParsers)) {
+        // known parser type
+        return parserPair->first(node[parserPair->second], rng);
+    }
+    return std::make_unique<ConstantAppender<double>>(node.to<double>());
+}
+/**
+ * @param node
+ *   a top-level document value i.e. either a scalar or a `^String` value
+ * @return
+ *   either a `^String` generator (etc--see `intParsers`)
+ *   or a constant generator if given a constant/scalar.
+ */
+UniqueGenerator<std::string> stringGenerator(const Node& node, DefaultRandom& rng) {
+    // Set of parsers to look when we request an int parser
+    // see int64Generator
+    const static std::map<std::string, Parser<UniqueGenerator<std::string>>> stringParsers{
+        {"^FastRandomString",
+         [](const Node& node, DefaultRandom& rng) {
+             return std::make_unique<FastRandomStringGenerator>(node, rng);
+         }},
+        {"^RandomString",
+         [](const Node& node, DefaultRandom& rng) {
+             return std::make_unique<NormalRandomStringGenerator>(node, rng);
+         }},
+        {"^Join",
+         [](const Node& node, DefaultRandom& rng) {
+             return std::make_unique<JoinGenerator>(node, rng);
+         }},
+        {"^Choose",
+         [](const Node& node, DefaultRandom& rng) {
+             return std::make_unique<ChooseStringGenerator>(node, rng);
+         }},
+        {"^IP",
+         [](const Node& node, DefaultRandom& rng) {
+             return std::make_unique<IPGenerator>(node, rng);
+         }},
+    };
+
+    if (auto parserPair = extractKnownParser(node, rng, stringParsers)) {
+        // known parser type
+        return parserPair->first(node[parserPair->second], rng);
+    }
+    return std::make_unique<ConstantAppender<std::string>>(node.to<std::string>());
+}
+
+ChooseGenerator::ChooseGenerator(const Node& node, DefaultRandom& rng) : _rng{rng} {
+    if (!node["from"].isSequence()) {
+        std::stringstream msg;
+        msg << "Malformed node for choose from array. Not a sequence " << node;
+        BOOST_THROW_EXCEPTION(InvalidValueGeneratorSyntax(msg.str()));
+    }
+    for (const auto&& [k, v] : node["from"]) {
+        _choices.push_back(valueGenerator<true, UniqueAppendable>(v, rng, allParsers));
+    }
+    if (node["weights"]) {
+        for (const auto&& [k, v] : node["weights"]) {
+            _weights.push_back(v.to<int64_t>());
+        }
+    } else {
+        // If not passed in, give each choice equal weight
+        _weights.assign(_choices.size(), 1);
+    }
```

Comment:
there's a bit of DRYing that could be done with this and the string choose logic. Not sure if worth it but at least a comment in both places would help if we have to fix/change in the future.